homeassistant/components/ecobee/__init__.py

```diff
@@ -1,123 +1,167 @@
 
-def setup_ecobee(hass, network, config):
-    """Set up the Ecobee thermostat."""
-    # If ecobee has a PIN then it needs to be configured.
-    if network.pin is not None:
-        request_configuration(network, hass, config)
-        return
+    if not await hass.async_add_executor_job(data.refresh):
+        return False
 
-    if "ecobee" in _CONFIGURING:
-        configurator = hass.components.configurator
-        configurator.request_done(_CONFIGURING.pop("ecobee"))
+    await hass.async_add_executor_job(data.update)
 
-    hold_temp = config[DOMAIN].get(CONF_HOLD_TEMP)
+    if data.ecobee.thermostats is None:
+        _LOGGER.error("No ecobee devices found to set up.")
+        return False
 
-    discovery.load_platform(hass, "climate", DOMAIN, {"hold_temp": hold_temp}, config)
-    discovery.load_platform(hass, "sensor", DOMAIN, {}, config)
-    discovery.load_platform(hass, "binary_sensor", DOMAIN, {}, config)
-    discovery.load_platform(hass, "weather", DOMAIN, {}, config)
+    hass.data[DOMAIN] = data
+
+    for component in ECOBEE_PLATFORMS:
+        hass.async_create_task(
+            hass.config_entries.async_forward_entry_setup(entry, component)
+        )
+
+    return True
 
 
 class EcobeeData:
-    """Get the latest data and update the states."""
+    """
+    Handle getting the latest data from ecobee.com so platforms can use it.
 
-    def __init__(self, config_file):
-        """Init the Ecobee data object."""
-        from pyecobee import Ecobee
+    Also handle refreshing tokens and updating config entry with refreshed tokens.
+    """
 
-        self.ecobee = Ecobee(config_file)
+    def __init__(self, hass, entry, api_key, refresh_token):
+        """Initialize the Ecobee data object."""
+        from pyecobee import Ecobee, ECOBEE_API_KEY, ECOBEE_REFRESH_TOKEN
+
+        self._hass = hass
+        self._entry = entry
+        self.ecobee = Ecobee(
+            config={ECOBEE_API_KEY: api_key, ECOBEE_REFRESH_TOKEN: refresh_token}
+        )
 
     @Throttle(MIN_TIME_BETWEEN_UPDATES)
     def update(self):
-        """Get the latest data from pyecobee."""
-        self.ecobee.update()
-        _LOGGER.debug("Ecobee data updated successfully")
-
-
-def setup(hass, config):
-    """Set up the Ecobee.
+        """Get the latest data from ecobee.com."""
+        from pyecobee import ExpiredTokenError
+
+        try:
+            self.ecobee.update()
+            _LOGGER.debug("Updating ecobee.")
+        except ExpiredTokenError:
+            _LOGGER.warn("ecobee update failed; attempting to refresh expired tokens.")
+            self.refresh()
+
+    def refresh(self) -> bool:
+        """Refresh ecobee tokens and update config entry."""
+        from pyecobee import ECOBEE_API_KEY, ECOBEE_REFRESH_TOKEN
+
+        _LOGGER.debug("Refreshing ecobee tokens and updating config entry.")
+        if self.ecobee.refresh_tokens():
+            self._hass.config_entries.async_update_entry(
```

Comment:
Is the fix to wrap async_update_entry with asyncio.run_coroutine_threadsafe?

Reply:
I would probably make all methods in `EcobeeData` coroutines and schedule the sync calls on the executor in those methods.

Since it's only two sync calls to update the api, and the results are cached, we can easily convert the `update` method in the entities to `async_update`.

Reply:
How do the actuator method in the entities work? Do they do I/O?

Reply:
The entities all call EcobeeData.update() in their own update methods, so those would need to be converted to async_update if EcobeeData.update was changed to be a coroutine. The climate platform also has several calls in various methods to other aspects of pyecobee that are also doing I/O in order to change thermostat settings.

Reply:
Ok. My suggestion should work then. We just convert the entity update methods to async_update coroutines and convert the update and refresh methods in `EcobeeData` to coroutines. Schedule the api calls in those two methods on the executor.

Actuator methods are left as is.